test/kpro_fetch_tests.erl

```diff
@@ -81,21 +88,45 @@ drop_older_offsets(Offset, [#kafka_message{offset = O} | R] = ML) ->
     false -> ML
   end.
 
-validate_messages([], Rest) -> Rest;
-validate_messages([#kafka_message{key = K, value = V} | R1], [Msg | R2]) ->
-  ok = validate_message(K, V, Msg),
-  validate_messages(R1, R2).
+validate_messages(_Topic, _FetchVsn, [], Rest) -> Rest;
+validate_messages(Topic, FetchVsn, [#kafka_message{ts_type = TType, ts = T, key = K, value = V} | R1],
+                  [{ProduceVsn, LAT, Msg} | R2]) ->
+  Magic = message_magic(FetchVsn, ProduceVsn),
+  ok =
+    case ?TOPIC_LAT == Topic of
+      true -> validate_message_append(Magic, TType, T, K, V, LAT, Msg);
+      false -> validate_message_create(Magic, TType, T, K, V, Msg)
+    end,
+  validate_messages(Topic, FetchVsn, R1, R2).
 
-validate_message(K, V, {K, V}) -> ok;
-validate_message(K, V, {_T, K, V}) -> ok;
-validate_message(K, V, #{key := K, value := V}) -> ok;
-validate_message(K, V, Wat) ->
-  erlang:error(#{ fetched => {K, V}
+validate_message_create(_Magic = 0, _TType, _T, K, V, #{key := K, value := V}) -> ok;
+validate_message_create(_Magic, create, T, K, V, #{ts := T, key := K, value := V}) -> ok;
+validate_message_create(Magic, TType, T, K, V, Wat) ->
+  erlang:error(#{ fetched => {T, K, V}
                 , produced => Wat
+                , magic => Magic
+                , ts_type => TType
                 }).
 
-do_fetch(Connection, Vsn, BeginOffset, MaxBytes) ->
-  Req = make_req(Vsn, BeginOffset, MaxBytes),
+validate_message_append(_Magic = 0, _TType, _T, K, V, _LAT, #{key := K, value := V}) -> ok;
+validate_message_append(_Magic, append, TF, K, V, undefined, #{ts := TP, key := K, value := V}) when TF /= TP -> ok;
+validate_message_append(_Magic, append, T, K, V, T, #{ts := TP, key := K, value := V}) when T /= TP -> ok;
+validate_message_append(Magic, TType, T, K, V, LAT, Wat) ->
+  erlang:error(#{ fetched => {T, K, V}
+                , produced => Wat
+                , magic => Magic
+                , ts_type => TType
+                , log_append_time => LAT
+                }).
```

Comment:
This should be a bit different from the `validate_message_create`, because if the timestamp is present depends solely on the fetch api version. Also it depends on the producer api version used if we do or do not have the LAT available.

So the clauses could be as follows:
* `FetchVsn < 2` → no timestamp check
* `ProduceVsn < 2` → no LAT, check only produce timestamp differs from the fetched timestamp
* Anything else → check both LAT and produce/fetch timestamp difference.

Another idea is to check that `LogAppendTime` is returned in `produce_randomly` when the produce api vsn is >= 2. The `validate_message_append` will be just slightly changed and the `Magic` would depend just on the fetch api version.

It's getting more and more complicated, I will think about how to make the tests more readable.

Reply:
To simplify this a lot, it could go in the opposite direction - assert based just on the timestamp type in the fetched messages (and don't care about the api versions during assertions).

The `validate_message` then could look like this:
```erlang
% message magic 0, no timestamp
validate_message(undefined, _T, K, V, #{key := K, value := V}) -> ok;
% message produced without timestamp
validate_message(create, -1, K, V, #{key := K, value := V}) -> ok;
% message produced with matching create timestamp
validate_message(create, T, K, V, #{ts := T, key := K, value := V}) -> ok;
% message produced with create timestamp which differs from the fetched log append timestamp
validate_message(append, TF, K, V, #{ts := TP, key := K, value := V}) when TF /= TP -> ok;
% something unexpected
validate_message(TType, T, K, V, Wat) ->
  erlang:error(#{ fetched => {TType, T, K, V}
                , produced => Wat
                }).
```

Reply:
I've pushed the simple approach here: https://github.com/hrubi/kafka_protocol/commits/decode-log-append-time-simple-tests

@k32 Thanks for taking care of it. Please let me know what do you think about the tests.